docs/HowToFAQ/Errors-Compiler.md

```diff
@@ -1,104 +1,529 @@
+If the assumption marked with `{:axiom}` is not actually valid, then the validity of the entire program is in question.
 
-## nondeterministic assignment forbidden by the --enforce-determinism option
+## **Error: a forall statement without a body cannot be compiled**
 
-## assign-such-that statement forbidden by the --enforce-determinism option
+<!-- %check-run -->
+```dafny
+method m(a: array<int>) {
+  forall x: int | x > 0 
+  var y := 3;
+}
+```
 
-## this assign-such-that statement is too advanced for the current compiler; Dafny's heuristics cannot find any bound for variable '{0}'
+A method may be parsed and verified even if a [forall statement](../DafnyRef/DafnyRef#sec-forall-statement) is missing a body. 
+However, the body must be supplied before the program can be compiled,
+even if the method is `ghost`. Body-less foralls in ghost methods are 
+similar to unchecked assumptions.
+
+## **Error: a loop without a body cannot be compiled**
+
+<!-- %check-run -->
+```dafny
+ghost method m(i: int) {
+  var x: int := i;
+  while x > 0 
+  x := 3;
+}
+```
 
-## nondeterministic if statement forbidden by the --enforce-determinism option
+A method may be parsed and verified even if a loop is missing a body. 
+Note that a missing body is different than an empty body, which is just `{ }`.
+However, the body must be supplied before the program can be compiled,
+even if the method is `ghost`. 
+Body-less loops in ghost methods are similar to unchecked assumptions.
 
-## binding if statement forbidden by the --enforce-determinism option
+## **Error: nondeterministic assignment forbidden by the --enforce-determinism option**
 
-## case-based if statement forbidden by the --enforce-determinism option
+<!-- %check-run %options --enforce-determinism -->
+```dafny
+method m() {
+  var x: int;
+  x := *;
+}
+```
+
+The `--enforce-determinism` option requires a target program to be deterministic and predictable:
+there may be no program statements that have an arbitrary, even if deterministic, result.
+Hence this 'assign any legal value' (`... := *`) statement is not permitted with `--enforce-determinism`.
+
+There are a few different forms of this kind of assignment:
+- `x := *;`
+- `x, y, z := 3, *, 42;`
+- `forall i | 0 <= i < a.Length { a[i] := *; }`
+
+## **Error: assign-such-that statement forbidden by the --enforce-determinism option**
+
+<!-- %check-run %options --enforce-determinism -->
+```dafny
+method m() {
+  var x: int;
+  x :| x < 5;
+}
+```
 
-## nondeterministic loop forbidden by the --enforce-determinism option
+The `--enforce-determinism` option requires a target program to be deterministic and predictable:
+there may be no program statements that have an arbitrary, even if deterministic, result.
+Hence this 'assign any value that satisfies the predicate' (`:|`) statement is not permitted with `--enforce-determinism`,
+even if there is only one such possible value.
```

Comment:
I could imagine someone saying "why?" about "even if there is only one such possible value". Is it worth explaining here?

Reply:
I added an explanatory sentence -- essentially it might be too hard to do and it might be too hard to determine whether it is too hard to do.